File: recordOfProblems/c++/ValidParenthesis.cc
bool areOpposite(char a, char b) {
    switch(a) {
        case '(': return b == ')'; break;
        case '{': return b == '}'; break;
        case '[': return b == ']'; break;
        default: return false;
    }
    return false;
}

bool isValid(string s) {
    std::stack<char> stack;
    
    for(int i=0; i<s.length(); i++) {
        if(!stack.empty()) {
            char topChar = stack.top();
            if(areOpposite(topChar, s[i])) {
                stack.pop();
            }
            else {
                stack.push(s[i]);
            }
        }
        else {
            stack.push(s[i]);
        }
    }
    
    return stack.empty();
}

File: Blind75Problems/Strings/ValidParenthesis.java
import java.util.Stack;

public class ValidParenthesis {

    public boolean areOpposite(char a, char b) {
        switch(a) {
            case '(': return b == ')';
            case '[': return b == ']';
            case '{': return b == '}';
            default:
        }
        return false;
    }

    public boolean isValidParenthesis(String s) {
        Stack<Character> stack = new Stack<Character>();

        for(char c : s.toCharArray()) {
            if(stack.isEmpty()) {
                stack.push(c);
            }
            else {
                char top = stack.peek();
                if(areOpposite(top, c)) {
                    stack.pop();
                }
                else {
                    stack.push(c);
                }
            }
        }
        return stack.isEmpty();
    }

    public void run() {
        // String s = "()[]{}";
        String s = "([]){}";
        System.out.println(isValidParenthesis(s));
    }
    public static void main(String[] args) {
        new ValidParenthesis().run();
    }
}